compiler/rustc_mir_transform/src/coverage/mod.rs

```diff
@@ -272,20 +281,54 @@ fn make_code_region(
         source_map.span_to_diagnostic_string(body_span)
     );
 
-    let (file, mut start_line, mut start_col, mut end_line, mut end_col) =
-        source_map.span_to_location_info(span);
-    if span.hi() == span.lo() {
-        // Extend an empty span by one character so the region will be counted.
-        if span.hi() == body_span.hi() {
-            start_col = start_col.saturating_sub(1);
-        } else {
-            end_col = start_col + 1;
-        }
+    let lo = span.lo();
+    let hi = span.hi();
+
+    let file = source_map.lookup_source_file(lo);
+    if !file.contains(hi) {
+        debug!(?span, ?file, ?lo, ?hi, "span crosses multiple files; skipping");
+        return None;
+    }
+
+    // Column numbers need to be in bytes, so we can't use the more convenient
+    // `SourceMap` methods for looking up file coordinates.
+    let rpos_and_line_and_byte_column = |pos: BytePos| -> Option<(RelativeBytePos, usize, usize)> {
+        let rpos = file.relative_position(pos);
+        let line_index = file.lookup_line(rpos)?;
+        let line_start = file.lines()[line_index];
+        // Line numbers and column numbers are 1-based, so add 1 to each.
+        Some((rpos, line_index + 1, (rpos - line_start).to_usize() + 1))
     };
-    if let Some(file) = file {
-        start_line = source_map.doctest_offset_line(&file.name, start_line);
-        end_line = source_map.doctest_offset_line(&file.name, end_line);
+
+    let (lo_rpos, mut start_line, mut start_col) = rpos_and_line_and_byte_column(lo)?;
+    let (hi_rpos, mut end_line, mut end_col) = rpos_and_line_and_byte_column(hi)?;
+
+    // If the span is empty, try to expand it by one character so that it is
+    // more visible in `llvm-cov` reports. (These column numbers are bytes, so
+    // "one character" might be multiple bytes.)
```

Comment:
it may be simpler to do this before computing the byte positions by modifying `hi` and `lo` right after computing them

Reply:
The subtlety here is that in the very common case of an empty span appearing at the end of a line, we still want to increase the column number by 1 (into a fake column) rather than wrapping to the start of the next line.

That way the empty span becomes a clearly visible single-space span at the end of the line. If we adjusted before finding line/column coordinates, the adjusted span would stretch from the end of one line to the start of the next, which ends up being invisible in coverage reports.

Reply:
(I discovered this the hard way, so it's probably worth documenting explicitly in the source.)

Reply:
The fact that we do this nudge at all is kind of annoying, but it’s so helpful in practice that we can’t really contemplate getting rid of it until we have branch coverage support as a semi-replacement.